Content.Shared/DoAfter/SharedDoAfterSystem.cs

```diff
@@ -392,5 +392,39 @@ public bool IsRunning(EntityUid entity, ushort id, DoAfterComponent? comp = null
     {
         return GetStatus(entity, id, comp) == DoAfterStatus.Running;
     }
+
+    /// <summary>
+    /// Returns a distinct list of entities which are currently interacting with the specified target entity.
+    /// This is done by finding the <see cref="DoAfterArgs.User"/> for each currently ongoing DoAfter with the given <see cref="DoAfterArgs.Target"/>.
+    /// </summary>
+    public List<EntityUid> GetEntitiesInteractingWithTarget(EntityUid target)
+    {
+        HashSet<EntityUid> result = new();
+        var doAfters = Array.Empty<DoAfter>();
+
+        var enumerator = EntityQueryEnumerator<ActiveDoAfterComponent, DoAfterComponent>();
+        while (enumerator.MoveNext(out var uid, out var active, out var comp))
+        {
+            var values = comp.DoAfters.Values;
+            var count = values.Count;
+
+            if (doAfters.Length < count)
+                doAfters = new DoAfter[count];
+
+            values.CopyTo(doAfters, 0);
+
+            for (var i = 0; i < count; i++)
+            {
+                var doAfter = doAfters[i];
+                if (doAfter.Cancelled || doAfter.Completed)
+                    continue;
+
+                if (doAfter.Args.Target == target)
+                    result.Add(doAfter.Args.User);
+            }
+        }
+
+        return new List<EntityUid>(result);
+    }
     #endregion
```

Comment:
1. Should have caller pass the list in and leave memory management on them.
2. Just have the caller pass in a hashset instead I would think.
3. Should this also consider BUIs? Consider someone stripping someone else for example.
4. I don't think you should need to copy the doafters as they're not being mutated here.
5. This should probably be on interactionsystem somewhere instead.

Reply:
> Should this also consider BUIs? Consider someone stripping someone else for example.

I know nothing about BUIs, but in the context of stripping specifically I don't think the UI needs to be considered since the actual action of stripping is a DoAfter.

The other changes are hopefully all implemented now.